File: apps/server/src/http/public/devices.ts
import type { FastifyInstance } from 'fastify'
import type Database from 'better-sqlite3'
import { z } from 'zod'

type DeviceRow = {
  pubkey: Buffer
  mac: Buffer
  owner_pubkey: Buffer
  name: string
  type: number
  location: number | null
  status: number
  hardware: string
  created_at: number
}

type TelemetryRow = {
  seq: number
  timestamp: number
  received_at: number
  ph_plain: number | null;           ph_cipher: Buffer | null
  ec_plain: number | null;           ec_cipher: Buffer | null
  water_level_plain: number | null;  water_level_cipher: Buffer | null
  temp_water_plain: number | null;   temp_water_cipher: Buffer | null
  temp_ambient_plain: number | null; temp_ambient_cipher: Buffer | null
  humidity_plain: number | null;     humidity_cipher: Buffer | null
}

function formatField(plain: number | null, cipher: Buffer | null) {
  if (plain !== null) return { value: plain }
  if (cipher !== null) return { encrypted: cipher.toString('hex') }
  return null
}

function formatDevice(r: DeviceRow) {
  return {
    id:          r.pubkey.toString('hex'),
    mac:         r.mac.toString('hex'),
    ownerPubkey: r.owner_pubkey.toString('hex'),
    name:        r.name,
    type:        r.type,
    location:    r.location,
    status:      r.status,
    hardware:    JSON.parse(r.hardware) as Record<string, string>,
    createdAt:   r.created_at,
  }
}

const FieldPolicySchema = z.object({
  default_disposition: z.number().int().min(0).max(2),
  per_destination:     z.record(z.number().int().min(0).max(2)).default({}),
})

const RegisterDeviceBody = z.object({
  id:            z.string().length(64),
  mac:           z.string().length(12),
  ownerPubkey:   z.string().length(64),
  ownerName:     z.string().optional(),
  name:          z.string().min(1),
  type:          z.number().int().min(0).max(2).default(0),
  publishTo:     z.number().int().min(0).max(2).default(1),
  location:      z.number().optional(),
  networks:      z.array(z.string()).default([]),
  localServers:  z.array(z.string()).default([]),
  privacyPolicy: z.object({
    ph:          FieldPolicySchema.optional(),
    ec:          FieldPolicySchema.optional(),
    water_level: FieldPolicySchema.optional(),
    temp_water:  FieldPolicySchema.optional(),
    temp_ambient:FieldPolicySchema.optional(),
    humidity:    FieldPolicySchema.optional(),
  }).optional(),
  hardware: z.object({
    model:            z.string(),
    firmware_version: z.string(),
  }).optional(),
})

const DEFAULT_PRIVACY_POLICY = {
  ph:          { default_disposition: 1, per_destination: {} },
  ec:          { default_disposition: 1, per_destination: {} },
  water_level: { default_disposition: 1, per_destination: {} },
  temp_water:  { default_disposition: 1, per_destination: {} },
  temp_ambient:{ default_disposition: 1, per_destination: {} },
  humidity:    { default_disposition: 1, per_destination: {} },
}

export async function registerDeviceRoutes(
  app: FastifyInstance,
  publicDb: Database.Database,
): Promise<void> {
  app.post('/v1/devices', async (request, reply) => {
    const result = RegisterDeviceBody.safeParse(request.body)
    if (!result.success) {
      return reply.code(400).send({ error: 'validation_error', details: result.error.issues })
    }

    const b = result.data
    const devicePubkey = Buffer.from(b.id, 'hex')
    const ownerPubkey  = Buffer.from(b.ownerPubkey, 'hex')
    const mac          = Buffer.from(b.mac, 'hex')

    const existing = publicDb
      .prepare('SELECT pubkey FROM devices WHERE pubkey = ?')
      .get(devicePubkey)
    if (existing) {
      return reply.code(409).send({ error: 'device_already_exists' })
    }

    // upsert owner — allows registering a device without a prior user creation step
    publicDb.prepare(`
      INSERT INTO users (pubkey, name, created_at)
      VALUES (?, ?, ?)
      ON CONFLICT (pubkey) DO NOTHING
    `).run(ownerPubkey, b.ownerName ?? b.ownerPubkey.slice(0, 12), Date.now())

    const now = Date.now()
    publicDb.prepare(`
      INSERT INTO devices
        (pubkey, mac, owner_pubkey, name, type, publish_to, location,
         networks, local_servers, privacy_policy, hardware, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(
      devicePubkey,
      mac,
      ownerPubkey,
      b.name,
      b.type,
      b.publishTo,
      b.location ?? null,
      JSON.stringify(b.networks),
      JSON.stringify(b.localServers),
      JSON.stringify(b.privacyPolicy ?? DEFAULT_PRIVACY_POLICY),
      JSON.stringify(b.hardware ?? {}),
      now,
    )

    const row = publicDb
      .prepare('SELECT pubkey, mac, owner_pubkey, name, type, location, status, hardware, created_at FROM devices WHERE pubkey = ?')
      .get(devicePubkey) as DeviceRow

    return reply.code(201).send({ device: formatDevice(row) })
  })

  app.get('/v1/devices', async (_request, reply) => {
    const rows = publicDb
      .prepare(`
        SELECT pubkey, mac, owner_pubkey, name, type, location, status, hardware, created_at
        FROM devices
      `)
      .all() as DeviceRow[]

    return reply.send({ devices: rows.map(formatDevice) })
  })

  app.get<{ Params: { id: string } }>('/v1/devices/:id', async (request, reply) => {
    const pubkey = Buffer.from(request.params.id, 'hex')
    const row = publicDb
      .prepare(`
        SELECT pubkey, mac, owner_pubkey, name, type, location, status, hardware, created_at
        FROM devices WHERE pubkey = ?
      `)
      .get(pubkey) as DeviceRow | undefined

    if (!row) return reply.code(404).send({ error: 'Device not found' })
    return reply.send({ device: formatDevice(row) })
  })

  app.get<{ Params: { id: string } }>('/v1/devices/:id/telemetry', async (request, reply) => {
    const pubkey = Buffer.from(request.params.id, 'hex')

    const rows = publicDb
      .prepare(`
        SELECT seq, timestamp, received_at,
          ph_plain, ph_cipher,
          ec_plain, ec_cipher,
          water_level_plain, water_level_cipher,
          temp_water_plain, temp_water_cipher,
          temp_ambient_plain, temp_ambient_cipher,
          humidity_plain, humidity_cipher
        FROM telemetry
        WHERE device_pubkey = ?
        ORDER BY timestamp DESC
        LIMIT 500
      `)
      .all(pubkey) as TelemetryRow[]

    return reply.send({
      readings: rows.map((r) => ({
        seq:         r.seq,
        timestamp:   r.timestamp,
        receivedAt:  r.received_at,
        ph:          formatField(r.ph_plain, r.ph_cipher),
        ec:          formatField(r.ec_plain, r.ec_cipher),
        waterLevel:  formatField(r.water_level_plain, r.water_level_cipher),
        tempWater:   formatField(r.temp_water_plain, r.temp_water_cipher),
        tempAmbient: formatField(r.temp_ambient_plain, r.temp_ambient_cipher),
        humidity:    formatField(r.humidity_plain, r.humidity_cipher),
      })),
    })
  })
}

File: apps/raiznetd/src/http/devices.rs
// Rotas de devices — paridade com apps/server/src/http/public/devices.ts.
//
// As validações replicam o zod do TS; em falha, 400 {"error":"validation_error",
// "details":[...]} (o corpus valida só status e a chave error; o shape de
// details pode divergir do zod).

use super::AppState;
use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use serde::Deserialize;
use serde_json::{Value, json};

#[derive(Deserialize)]
pub struct FieldPolicyIn {
    pub default_disposition: u8,
    #[serde(default)]
    pub per_destination: std::collections::HashMap<String, u8>,
}

/// Body do POST /v1/devices. rename_all = "camelCase" mapeia owner_pubkey ↔
/// "ownerPubkey" etc. `#[serde(default)]` aplica o default se o campo faltar.
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegisterBody {
    pub id: String,
    pub mac: String,
    pub owner_pubkey: String,
    pub owner_name: Option<String>,
    pub name: String,
    #[serde(default)]
    pub r#type: i64, // `type` é palavra reservada; r# permite usá-la como nome
    #[serde(default = "default_publish_to")]
    pub publish_to: i64,
    pub location: Option<i64>,
    #[serde(default)]
    pub networks: Vec<String>,
    #[serde(default)]
    pub local_servers: Vec<String>,
    pub privacy_policy: Option<std::collections::HashMap<String, FieldPolicyIn>>,
    pub hardware: Option<Value>,
}

fn default_publish_to() -> i64 {
    1
}

/// Política default quando o body não traz nenhuma: tudo plain (disposition 1)
/// — igual a DEFAULT_PRIVACY_POLICY do TS.
fn default_privacy_policy() -> Value {
    let field = json!({ "default_disposition": 1, "per_destination": {} });
    json!({
        "ph": field, "ec": field, "water_level": field,
        "temp_water": field, "temp_ambient": field, "humidity": field
    })
}

/// Shape do device nas respostas — igual ao formatDevice do TS.
fn format_device(d: &raiznet_store::devices::DeviceRow) -> Value {
    json!({
        "id": hex::encode(&d.pubkey),
        "mac": hex::encode(&d.mac),
        "ownerPubkey": hex::encode(&d.owner_pubkey),
        "name": d.name,
        "type": d.device_type,
        "location": d.location,
        "status": d.status,
        "hardware": serde_json::from_str::<Value>(&d.hardware).unwrap_or_else(|_| json!({})),
        "createdAt": d.created_at,
    })
}

fn validation_error(details: Vec<String>) -> (StatusCode, Json<Value>) {
    (
        StatusCode::BAD_REQUEST,
        Json(json!({ "error": "validation_error", "details": details })),
    )
}

pub async fn register(
    State(state): State<AppState>,
    // Recebemos Value e fazemos o parse manualmente para que um body
    // malformado vire 400 validation_error (como o zod do TS), e não o
    // 422 padrão do extractor Json<T> do axum.
    Json(value): Json<Value>,
) -> (StatusCode, Json<Value>) {
    let body: RegisterBody = match serde_json::from_value(value) {
        Ok(b) => b,
        Err(e) => return validation_error(vec![e.to_string()]),
    };

    let mut details = Vec::new();
    // unwrap_or_default: hex inválido vira vetor vazio, que reprova no
    // check de tamanho logo abaixo.
    let pubkey = hex::decode(&body.id).unwrap_or_default();
    let mac = hex::decode(&body.mac).unwrap_or_default();
    let owner = hex::decode(&body.owner_pubkey).unwrap_or_default();
    if body.id.len() != 64 || pubkey.len() != 32 {
        details.push("id must be 64 hex chars".into());
    }
    if body.mac.len() != 12 || mac.len() != 6 {
        details.push("mac must be 12 hex chars".into());
    }
    if body.owner_pubkey.len() != 64 || owner.len() != 32 {
        details.push("ownerPubkey must be 64 hex chars".into());
    }
    if body.name.is_empty() {
        details.push("name must not be empty".into());
    }
    if !(0..=2).contains(&body.r#type) {
        details.push("type must be 0..2".into());
    }
    if !(0..=2).contains(&body.publish_to) {
        details.push("publishTo must be 0..2".into());
    }
    if let Some(pp) = &body.privacy_policy {
        if pp
            .values()
            .any(|f| f.default_disposition > 2 || f.per_destination.values().any(|d| *d > 2))
        {
            details.push("disposition must be 0..2".into());
        }
    }
    if !details.is_empty() {
        return validation_error(details);
    }

    // Política: a do body (re-serializada) ou a default tudo-plain.
    let policy_json = body
        .privacy_policy
        .as_ref()
        .map(|pp| {
            serde_json::to_value(
                pp.iter()
                    .map(|(k, f)| {
                        (
                            k.clone(),
                            json!({
                                "default_disposition": f.default_disposition,
                                "per_destination": f.per_destination,
                            }),
                        )
                    })
                    .collect::<serde_json::Map<_, _>>(),
            )
            .expect("serializable")
        })
        .unwrap_or_else(default_privacy_policy);

    let db = state.devices_db().lock().expect("db mutex");
    match raiznet_store::devices::device_exists(&db, &pubkey) {
        Ok(true) => {
            // O firmware trata este 409 como sucesso (registro lazy).
            return (StatusCode::CONFLICT, Json(json!({ "error": "device_already_exists" })));
        }
        Ok(false) => {}
        Err(_) => {
            return (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({ "error": "internal" })));
        }
    }

    let now = crate::now_ms() as i64;
    // Owner sem nome vira o prefixo da pubkey — paridade com devices.ts:115.
    let owner_name = body
        .owner_name
        .clone()
        .unwrap_or_else(|| body.owner_pubkey[..12].to_string());
    let new_device = raiznet_store::devices::NewDevice {
        pubkey: &pubkey,
        mac: &mac,
        owner_pubkey: &owner,
        owner_name: &owner_name,
        name: &body.name,
        device_type: body.r#type,
        publish_to: body.publish_to,
        location: body.location,
        networks_json: &serde_json::to_string(&body.networks).expect("serializable"),
        local_servers_json: &serde_json::to_string(&body.local_servers).expect("serializable"),
        privacy_policy_json: &policy_json.to_string(),
        hardware_json: &body.hardware.clone().unwrap_or_else(|| json!({})).to_string(),
        created_at: now,
    };
    if raiznet_store::devices::insert_device(&db, &new_device).is_err() {
        return (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({ "error": "internal" })));
    }
    let row = raiznet_store::devices::get_device(&db, &pubkey)
        .ok()
        .flatten()
        .expect("just inserted");
    (StatusCode::CREATED, Json(json!({ "device": format_device(&row) })))
}

pub async fn list(State(state): State<AppState>) -> (StatusCode, Json<Value>) {
    let db = state.devices_db().lock().expect("db mutex");
    match raiznet_store::devices::list_devices(&db) {
        Ok(rows) => (
            StatusCode::OK,
            Json(json!({ "devices": rows.iter().map(format_device).collect::<Vec<_>>() })),
        ),
        Err(_) => (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({ "error": "internal" }))),
    }
}

pub async fn get_one(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> (StatusCode, Json<Value>) {
    let pubkey = hex::decode(&id).unwrap_or_default();
    let db = state.devices_db().lock().expect("db mutex");
    match raiznet_store::devices::get_device(&db, &pubkey) {
        Ok(Some(row)) => (StatusCode::OK, Json(json!({ "device": format_device(&row) }))),
        // Mensagem com espaço e maiúscula mesmo — é o contrato do TS.
        Ok(None) => (StatusCode::NOT_FOUND, Json(json!({ "error": "Device not found" }))),
        Err(_) => (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({ "error": "internal" }))),
    }
}

/// Um campo na resposta de leitura: {"value":x} | {"encrypted":"hex"} | null.
fn format_field(c: &raiznet_store::telemetry::SensorColumns) -> Value {
    match (&c.plain, &c.cipher) {
        (Some(v), _) => json!({ "value": v }),
        (None, Some(b)) => json!({ "encrypted": hex::encode(b) }),
        (None, None) => Value::Null,
    }
}

pub async fn telemetry_history(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> (StatusCode, Json<Value>) {
    let pubkey = hex::decode(&id).unwrap_or_default();
    let db = state.devices_db().lock().expect("db mutex");
    match raiznet_store::telemetry::query_telemetry(&db, &pubkey) {
        Ok(rows) => {
            let readings: Vec<Value> = rows
                .iter()
                .map(|r| {
                    json!({
                        "seq": r.seq, "timestamp": r.timestamp, "receivedAt": r.received_at,
                        "ph": format_field(&r.ph), "ec": format_field(&r.ec),
                        "waterLevel": format_field(&r.water_level),
                        "tempWater": format_field(&r.temp_water),
                        "tempAmbient": format_field(&r.temp_ambient),
                        "humidity": format_field(&r.humidity),
                    })
                })
                .collect();
            (StatusCode::OK, Json(json!({ "readings": readings })))
        }
        Err(_) => (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({ "error": "internal" }))),
    }
}
